src/beast/beast/http/basic_parser.h

```diff
@@ -113,153 +133,405 @@ class basic_parser
+    std::size_t
+    write(ConstBufferSequence const& buffers,
+        error_code& ec);
 
     /** Called to indicate the end of file.
+        
         HTTP needs to know where the end of the stream is. For example,
         sometimes servers send responses without Content-Length and
         expect the client to consume input (for the body) until EOF.
         Callbacks and errors will still be processed as usual.
+        
         @note This is typically called when a socket read returns eof.
-        @return `true` if the message is complete.
+        
+        @throws boost::system::system_error Thrown on failure.
     */
-    error_code
-    write_eof();
+    void
+    write_eof()
+    {
+        error_code ec;
+        write_eof(ec);
+        if(ec)
+            throw boost::system::system_error{ec};
+    }
 
-protected:
-    /** Called once when a new message begins. */
-    virtual
+    /** Called to indicate the end of file.
+        
+        HTTP needs to know where the end of the stream is. For example,
+        sometimes servers send responses without Content-Length and
+        expect the client to consume input (for the body) until EOF.
+        Callbacks and errors will still be processed as usual.
+        
+        @note This is typically called when a socket read returns eof.
+        
+        @param ec Set to the error, if any error occurred.
+    */
     void
-    on_start() = 0;
+    write_eof(error_code& ec);
+
+private:
+    Derived&
+    impl()
+    {
+        return *static_cast<Derived*>(this);
+    }
+
+    template<class C>
+    class has_on_start_t
+    {
+        template<class T, class R =
+            decltype(std::declval<T>().on_start(), std::true_type{})>
+        static R check(int);
+        template <class>
+        static std::false_type check(...);
+        using type = decltype(check<C>(0));
+    public:
+        static bool const value = type::value;
+    };
+    template<class C>
+    using has_on_start =
+        std::bool_constant<has_on_start_t<C>::value>;
 
-    /** Called for each header field. */
-    virtual
     void
-    on_field (std::string const& field, std::string const& value) = 0;
+    call_on_start(std::true_type)
+    {
+        impl().on_start();
+    }
+
+    void
+    call_on_start(std::false_type)
+    {
+    }
+
+    template<class C>
+    class has_on_field_t
+    {
+        template<class T, class R =
+            decltype(std::declval<T>().on_field(
+                std::declval<std::string>(), std::declval<std::string>()),
+                    std::true_type{})>
+        static R check(int);
+        template <class>
+        static std::false_type check(...);
+        using type = decltype(check<C>(0));
+    public:
+        static bool const value = type::value;
+    };
+    template<class C>
+    using has_on_field =
+        std::bool_constant<has_on_field_t<C>::value>;
+
+    void
+    call_on_field(std::string field, std::string value,
```

Comment:
Param type should be `const&`, unless you expect `impl().on_field` to consume the parameters, in which case we should leave the param type as is and change the call to `impl().on_field(std::move(field), std::move(value))`